src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -334,6 +337,18 @@ RubyBinding binding(Frame callerFrame, Object self, Object[] rubyArgs, RootCallT
                 @Cached(
                         value = "getAdoptedNode(this).getEncapsulatingSourceSection()",
                         allowUncached = true, neverDefault = false) SourceSection sourceSection) {
+
+            final MaterializedFrame materializedCallerFrame = Truffle.getRuntime()
+                    .iterateFrames(f -> f.getFrame(FrameInstance.FrameAccess.MATERIALIZE).materialize(), 1);
```

Comment:
We definitely can't do `iterateFrames` in Kernel#binding.
`Kernel#binding` is very well optimized currently and `iterateFrames` would completely ruin its performance.

One way to detect C extension methods quickly is https://github.com/oracle/truffleruby/pull/3436#discussion_r1479705647.
But also I don't think this matters. We do have currently a Ruby frame for C extension methods, so it's only fair to return a Binding for that case.

The aim of this new exception is to not give an empty/unconnected binding for situations where there isn't a real binding.
For instance `method(:binding).call` is such a case, because Method#call is implemented in C/Java and so there is no way to get a proper Binding for Method#call, hence there it's clearer to raise an exception.
As mentioned in https://github.com/oracle/truffleruby/pull/3436#discussion_r1478757097, we should raise the new exception for `method(:binding).call` and add a spec for that.
And leave C extension methods alone, i.e. let them return a Binding as they do, because there a real Ruby frame in there.